### DISCUSSION TITLE
How to update user's groups via Admin REST API?

### DISCUSSION THREAD

**Discussion by thunderamur** [keycloak/keycloak#8552] [Q&A] [answered]
I found this method in [docs](https://www.keycloak.org/docs-api/15.0/rest-api/index.html#_users_resource).
`PUT /{realm}/users/{id}`
```json
{
    "groups": ["cc8cb705-b021-4a67-88b0-1e73a632fa63"]
}
```

But it do not work for groups as expected, it just have no effect.

**Comment by ert-git** [comment, marked as answer]
Have the same problem. But it seems API for groups is the same to adding roles (special requests instead of mapping inside User data). Methods `joinGroup(groupId)/leaveGroup(groupId)` work for me.
Join (add) group:
```json
PUT /{realm}/users/{userId}/groups/{groupId}
{
  "realm":"{realm}",
  "userId":"{userId}",
  "groupId":"{groupId}"
}
```
Leave (remove) group:
```
DELETE /{realm}/users/{userId}/groups/{groupId}
```

**Reply by thunderamur** [reply]
Thanks for your answer. I found this method too. But it allows you to add groups one at a time. According to the UserRepresentation, I expected that an update is possible as I showed above.

**Comment by ZaheerUdDeen** [comment]
it would had been  great, if it works same as createUser end point.

**Comment by tiagopsantos** [comment]
Exactly what i can see the in-spite of receiving the same input (UserRepresentation) the updateUser doesn't use the groups

https://github.com/keycloak/keycloak/blob/f69497eb28ab6c6595a4a335607d03b0c7ba229d/services/src/main/java/org/keycloak/services/resources/admin/UserResource.java#L162

https://github.com/keycloak/keycloak/blob/f69497eb28ab6c6595a4a335607d03b0c7ba229d/services/src/main/java/org/keycloak/services/resources/admin/UsersResource.java#L115
---->            RepresentationToModel.createGroups(rep, realm, user);


Unfortunatelly i'm getting many javax.persistence.OptimisticLockException: org.hibernate.exception.LockAcquisitionException:
in the createUser...

	at org.keycloak.models.jpa.UserAdapter.joinGroup(UserAdapter.java:394)
	at org.keycloak.models.utils.RepresentationToModel.createGroups(RepresentationToModel.java:1903)
	at org.keycloak.services.resources.admin.UsersResource.createUser(UsersResource.java:161)

EDIT - Confirmed that the work-around worked:
```
The work-around is to explicitly add sendStringParametersAsUnicode with value set to false for DB connection URL
jdbc:sqlserver://db.example.com:1433;DatabaseName=dballo19;sendStringParametersAsUnicode=false

```
https://github.com/keycloak/keycloak/issues/15060

**Comment by leup** [comment]
Hello! Do you have any plans to move forward with this? Thank you
